pkg/kine/drivers/generic/admission.go

```diff
@@ -38,33 +49,46 @@ func (p *allowAllPolicy) Admit(ctx context.Context, txName string) (func(), erro
 type limitPolicy struct {
 	maxConcurrentTxn int64
 	semaphore        *semaphore.Weighted
+	onlyWriteQueries bool
 }
 
-func newLimitPolicy(maxConcurrentTxn int64) *limitPolicy {
+func newLimitPolicy(onlyWriteQueries bool, maxConcurrentTxn int64) *limitPolicy {
 	return &limitPolicy{
 		maxConcurrentTxn: maxConcurrentTxn,
 		semaphore:        semaphore.NewWeighted(maxConcurrentTxn),
+		onlyWriteQueries: onlyWriteQueries,
 	}
+
+}
+
+func (p *limitPolicy) shouldCheck(txName string) bool {
+	_, isWriteQuery := writeQueries[txName]
+	return !p.onlyWriteQueries || isWriteQuery
 }
 
 func (p *limitPolicy) Admit(ctx context.Context, txName string) (func(), error) {
-	ok := p.semaphore.TryAcquire(1)
-	if !ok {
-		recordOpAdmissionControl(txName, statusDenied)
-		return func() {}, fmt.Errorf("number of concurrent database operations reached limit (%d)", p.maxConcurrentTxn)
+	if p.shouldCheck(txName) {
```

Comment:
With this, we will evaluate `shouldCheck` twice for each query. Once here, and once in the callback. This is not ideal, since this is supposed to be on the hot path.

Further, I am not in favor of a function `shouldCheck`, as this does not give the caller any info about what it actually checks.

Given that we only need to do this once, we could just do the bool expression, with a comment explaining it (if needed)

```go
// we need to lock the semaphore if ...
needSem := !onlyWriteQueries || isWriteQuery[tx_name]
```

Reply:
Done